### PR TITLE
Enhancements to Quick Add Personal Vocab Feature

### PR DESCRIPTION
**Pull request by Tanvez**
[ATGU-3145](https://jira.huit.harvard.edu/browse/ATGU-3145) and [ATGU-3161](https://jira.huit.harvard.edu/browse/ATGU-3161)
- added select options for related nodes in modal
- added test for django and vue
- added preferred language
- make node selection accessible via arrow keys

Screen Recording 2021-07-29 at 2 07 25 PM

Update:
- refactor lattice_nodes api to look at lemma_nodes when building the lattice_nodes list relating to a headword
- refactor regex to include trailing numbers in results
- updated vue select options for suggested gloss to strip out numbers from lattice_node labels
- updated django test in relation to updated api

### CODE REVIEW

**Review by dodget — APPROVED** [review]
Great work Vesna! I'd like to answer a couple of questions before merging, mainly around adding the preferred language to the user profile.

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py
Comment (dodget) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: This rogue migration could be the result of django drift. Probably nothing that you did.

File: hedera/api.py
Comment (Tanvez) [reply hedera/api.py:338]: oh I didn't know that! Thats a cool feature! Will update the code with those changes.

File: hedera/migrations/0004_profile_lang.py
Comment (Tanvez) [reply hedera/migrations/0004_profile_lang.py:18]: Yea, definitely! Will hold off on any merging until 👍 by Bill.

File: hedera/tests/test_api.py
Comment (Tanvez) [reply hedera/tests/test_api.py:69]: ahh will update the test with failing tests!

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py
Comment (Tanvez) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: It was generated when I ran the migrations for the Profile model, not sure if it was a previous change that was never migrated or django drift 🤔

File: hedera/tests/test_api.py
Comment (dodget) [inline hedera/tests/test_api.py:120]: 👍

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue
Comment (Tanvez) [inline static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:127]: todo: remove dead code

File: hedera/api.py
Comment (arthurian) [inline hedera/api.py:334]: Should we return these 400 responses as JSON, that way the client knows the errors will be returned as JSON as well? Maybe something like this (just a suggestion):

```python
class JsonResponseBadRequest(JsonResponse):
    status_code = 400
```

And then you could write:

```python
return JsonResponseBadRequest(data={"error": "Missing required fields"})
```

File: hedera/api.py
Comment (arthurian) [inline hedera/api.py:347]: Just curious, but what is `\y` in the regex?

File: hedera/migrations/0004_profile_lang.py
Comment (arthurian) [reply hedera/migrations/0004_profile_lang.py:18]: Agreed on getting input from @bilbe.

File: lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py
Comment (arthurian) [reply lemmatized_text/migrations/0015_alter_lemmatizedtext_cloned_for.py:20]: Yeah, I suspect this is drift. And as we discussed offline, we should pin it at django 3.2 now so that doesn't happen again (at least not without our knowing).

File: static/src/js/app/Dashboard.vue
Comment (arthurian) [inline static/src/js/app/Dashboard.vue:3]: I suggest we remove this, since we can assume it's under construction.

File: static/src/js/app/components/quick-add-button/QuickAddButton.vue
Comment (arthurian) [inline static/src/js/app/components/quick-add-button/QuickAddButton.vue:22]: I'm wondering if it would make sense to have the `FETCH_ME` action dispatched from the `Dashboard.vue` component instead of here. My assumption is that the "me" data only needs to be read once, when the dashboard is loaded. I could imagine other sub-components on the dashboard needing that data, and we wouldn't want every sub-component dispatching the same action. So if the dashboard managed that state by dispatching the action, then sub-components wouldn't need to manage it, they could just read the state directly from the store. I could be thinking about this wrong, so curious what you think.

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue
Comment (arthurian) [inline static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:221]: 👍  for the note about the `+`. That's a neat trick, although you could make the case that `parseInt()` is clearer and reflects the expected format of the number.

**Review by arthurian — APPROVED** [review]
Nice work, @Tanvez! It's great to see how the QuickAdd components are evolving, along with the backend and frontend unit tests. Kudos for making those better.

Overall this looks like it's in good shape to merge, other than a few questions around the preferred language that @dodget mentioned. I added a few questions/suggestions, but nothing that should block you from moving forward.

File: hedera/api.py
Comment (Tanvez) [reply hedera/api.py:334]: Yea, we definitely should! Will update the code to reflect this.

File: hedera/api.py
Comment (Tanvez) [reply hedera/api.py:347]: Think it might have something to do with how Postgres uses word boundaries. I found an old stackoverflow post here [PostgreSQL Regex Word Boundaries](https://stackoverflow.com/questions/3825676/postgresql-regex-word-boundaries/29798772). We might have to tweak this to accommodate the changes Ivy was talking about.

File: static/src/js/app/Dashboard.vue
Comment (Tanvez) [reply static/src/js/app/Dashboard.vue:3]: Will do!

File: static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue
Comment (Tanvez) [reply static/src/js/app/components/quick-add-button/QuickAddVocabForm.vue:221]: ahh that would be more descriptive! will update with the `parseInt()`

File: hedera/api.py
Comment (Tanvez) [reply hedera/api.py:347]: Should we hold off on refactoring the regex or refactor the regex to account for numerals in the headword?

File: static/src/js/app/components/quick-add-button/QuickAddButton.vue
Comment (Tanvez) [reply static/src/js/app/components/quick-add-button/QuickAddButton.vue]: I think due to how hedera loads in different vue components periodically some of the vuex state might not persist between page navigation. Since we are planning to add this button site wide, might need to be updated in [ATGU-3160](https://jira.huit.harvard.edu/browse/ATGU-3160)

Comment (arthurian) [reply static/src/js/app/components/quick-add-button/QuickAddButton.vue]: Yeah, it may require further refactoring when we get to that story.

File: hedera/api.py
Comment (Tanvez) [reply hedera/api.py:347]: @jtauber We had a question about the following when looking up headwords in relation to the personal vocabulary list:
- if we wanted to search for a headword should we look at `lattices_latticenodes`  table
- how are the headword stored based on multiple meanings between the same headword

File: hedera/api.py
Comment (Tanvez) [inline hedera/api.py:411]: todo: leave off check for valid_lemma

**Pull request merged**
